File: leetcode/rb/15.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  nums.sort!

  res = []

  (nums.size - 2).times do |i|
    next if i != 0 && nums[i] == nums[i - 1]

    start_i, end_i = i + 1, nums.size - 1

    while start_i < end_i
      total = nums[start_i] + nums[end_i] + nums[i]

      res.push([nums[i], nums[start_i], nums[end_i]]) if total == 0

      if total <= 0
        start_i += 1
        start_i += 1 while nums[start_i] == nums[start_i - 1] && start_i < end_i
      end

      if total >= 0
        end_i -= 1
        end_i -= 1 while nums[end_i] == nums[end_i + 1] && start_i < end_i
      end
    end
  end

  return res
end

File: leetcode/py/15.py
class Solution:
    def threeSum(self, nums):
        nums = sorted(nums)
        res = []

        for i, num in enumerate(nums):
            if i > 0 and num == nums[i - 1]:
                continue

            lo, hi = i + 1, len(nums) - 1
            while lo < hi:
                total = num + nums[lo] + nums[hi]
                if total < 0:
                    lo += 1
                elif total > 0:
                    hi -= 1
                else:
                    res.append([num, nums[lo], nums[hi]])
                    lo += 1
                    hi -= 1
                    while lo < hi and nums[lo] == nums[lo - 1]:
                        lo += 1
                    while lo < hi and nums[hi] == nums[hi + 1]:
                        hi -= 1

        return res
